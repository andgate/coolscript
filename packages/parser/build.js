const process = require('process')
const esbuild = require('esbuild')
const NearleyPlugin = require('esbuild-plugin-nearley')

let watch = process.argv.some((arg) => arg == '--watch')
if (watch) {
  watch = {
    onRebuild(error, result) {
      if (error) console.error('watch build failed:', error)
      else console.log('watch build succeeded:', result)
    }
  }
}

const shared = {
  entryPoints: ['src/index.ts'],
  outdir: 'lib/',
  target: 'es6',
  sourcemap: true,
  bundle: true,
  minify: false,
  tsconfig: 'tsconfig.build.json',
  plugins: [NearleyPlugin.default],
  watch
}

esbuild
  .build({
    ...shared,
    format: 'esm',
    splitting: true
  })
  .catch(() => process.exit(1))

esbuild
  .build({
    ...shared,
    format: 'cjs',
    outExtension: { '.js': '.cjs' }
  })
  .catch(() => process.exit(1))
