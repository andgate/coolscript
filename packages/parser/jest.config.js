const base = require('../../jest.config.base.js')
const { name: packageName } = require('./package.json')

module.exports = {
  ...base,
  name: packageName,
  displayName: packageName,
  moduleNameMapper: {
    '^@coolscript/parser$': '<rootDir>/src'
  },
  transform: {
    ...base.transform,
    '^.+\\.ne$': 'jest-transform-nearley'
  }
}
