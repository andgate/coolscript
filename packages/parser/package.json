{
  "name": "@coolscript/parser",
  "version": "0.0.3",
  "description": "Coolscript programming language parser.",
  "main": "./lib/index.js",
  "module": "./lib/index.cjs",
  "types": "./lib/index.d.ts",
  "publishConfig": {
    "main": "./lib/index.js",
    "module": "./lib/index.cjs",
    "typings": "lib/index.d.ts",
    "access": "public"
  },
  "files": [
    "lib/",
    "LICENSE"
  ],
  "scripts": {
    "start": "pnpm watch",
    "build": "pnpm build:lib",
    "build:lib": "pnpm clean:lib && pnpm build:grammar && pnpm build:ts && pnpm build:js",
    "build:js": "node build.js",
    "build:ts": "tsc",
    "build:grammar": "pnpm clean:grammar && nearleyc src/grammar.ne -o src/grammar.js",
    "watch": "concurrently \"pnpm:watch:*\"",
    "watch:js": "node build.js --watch",
    "watch:ts": "tsc --watch",
    "check": "tsc --noEmit",
    "prettier": "prettier --check src/ tests/",
    "prettier:fix": "prettier --write src/ tests/",
    "lint": "eslint src/** tests/** --ext .ts,.tsx",
    "lint:fix": "pnpm lint --fix",
    "verify": "pnpm prettier && pnpm lint",
    "verify:fix": "pnpm prettier:fix && pnpm lint:fix",
    "test": "jest",
    "clean": "pnpm clean:grammar && pnpm clean:lib && pnpm clean:etc",
    "clean:lib": "rimraf lib tsconfig.tsbuildinfo",
    "clean:grammar": "rimraf src/grammar.js",
    "clean:etc": "rimraf coverage .pnpm-debug.log tsconfig.tsbuildinfo",
    "clean:node_modules": "rimraf node_modules"
  },
  "author": "Gabriel Anderson",
  "license": "MIT",
  "devDependencies": {
    "@types/moo": "^0.5.5",
    "@types/nearley": "^2.11.2",
    "moo": "^0.5.1",
    "nearley": "^2.20.1"
  },
  "dependencies": {
    "@coolscript/syntax": "^0.0.3"
  },
  "gitHead": "9b91bf0df0bb73d74440b9a1d191fd5c91c6af93"
}
