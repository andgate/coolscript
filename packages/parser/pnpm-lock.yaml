lockfileVersion: 5.3

specifiers:
  parjs: ^0.12.7

dependencies:
  parjs: 0.12.7

packages:

  /char-info/0.3.2:
    resolution: {integrity: sha512-6P6KW8crZx+N857wZalB4FreR+PhheSLmAk22c8zbQsFhsDxZP1aTDfmOjrzddgp1IBOl53b0Z8kDQrwh4B//A==}
    dependencies:
      node-interval-tree: 1.3.3
    dev: false

  /lodash/4.17.21:
    resolution: {integrity: sha512-v2kDEe57lecTulaDIuNTPy3Ry4gLGJ6Z1O3vE1krgXZNrsQ+LFTGHVxVjcXPs17LhbZVGedAJv8XZ1tvj5FvSg==}
    dev: false

  /node-interval-tree/1.3.3:
    resolution: {integrity: sha512-K9vk96HdTK5fEipJwxSvIIqwTqr4e3HRJeJrNxBSeVMNSC/JWARRaX7etOLOuTmrRMeOI/K5TCJu3aWIwZiNTw==}
    engines: {node: '>= 7.6.0'}
    dependencies:
      shallowequal: 1.1.0
    dev: false

  /parjs/0.12.7:
    resolution: {integrity: sha512-aKw+vsSMBdZNdxcetIwi5mnmFckuwxTz7dKu07wEzrNjKYtvhRlrlla3b4GAEmGMyAWhrocLrUcbSGvmMZMU3Q==}
    dependencies:
      char-info: 0.3.2
      lodash: 4.17.21
    dev: false

  /shallowequal/1.1.0:
    resolution: {integrity: sha512-y0m1JoUZSlPAjXVtPPW70aZWfIL/dSP7AFkRnniLCrK/8MDKog3TySTBmckD+RObVxH0v4Tox67+F14PdED2oQ==}
    dev: false
