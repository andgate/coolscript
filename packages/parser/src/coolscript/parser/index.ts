export * from './coolscript-parser'
export * from './token'
