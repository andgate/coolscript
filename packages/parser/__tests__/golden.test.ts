import { examples } from '@coolscript/examples'
import { parse } from '@coolscript/parser'

const exampleKeys: string[] = Object.keys(examples)

describe('Parser Golden Test Suite', () => {
  test.each(exampleKeys)('Parse example %p', (exampleKey) => {
    const example = examples[exampleKey]
    const parseResult = parse(example)
    expect(parseResult).toMatchSnapshot(exampleKey)
  })
})
