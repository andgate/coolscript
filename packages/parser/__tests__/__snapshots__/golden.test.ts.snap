// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser Golden Test Suite Parse example "bool1": bool1 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 5,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "bool": true,
      "span": Object {
        "column": Object {
          "end": 5,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "BooleanValue",
    },
  },
  "warnings": Array [
    [Error: Ambiguous grammar encountered! Resolving first result.],
  ],
}
`;

exports[`Parser Golden Test Suite Parse example "bool2": bool2 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 6,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "bool": false,
      "span": Object {
        "column": Object {
          "end": 6,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "BooleanValue",
    },
  },
  "warnings": Array [
    [Error: Ambiguous grammar encountered! Resolving first result.],
  ],
}
`;

exports[`Parser Golden Test Suite Parse example "call1": call1 1`] = `
Object {
  "term": Object {
    "body": Object {
      "args": Array [
        Object {
          "span": Object {
            "column": Object {
              "end": 24,
              "start": 23,
            },
            "line": Object {
              "end": 1,
              "start": 1,
            },
            "type": "Span",
          },
          "tag": "ValueTerm",
          "value": Object {
            "num": 1,
            "span": Object {
              "column": Object {
                "end": 24,
                "start": 23,
              },
              "line": Object {
                "end": 1,
                "start": 1,
              },
              "type": "Span",
            },
            "tag": "NumberValue",
          },
        },
      ],
      "func": Object {
        "span": Object {
          "column": Object {
            "end": 22,
            "start": 21,
          },
          "line": Object {
            "end": 1,
            "start": 1,
          },
          "type": "Span",
        },
        "tag": "VariableTerm",
        "variable": "f",
      },
      "span": Object {
        "column": Object {
          "end": 25,
          "start": 21,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "CallTerm",
    },
    "declarations": Array [
      Object {
        "body": Object {
          "args": Array [
            "x",
          ],
          "body": Object {
            "span": Object {
              "column": Object {
                "end": 17,
                "start": 16,
              },
              "line": Object {
                "end": 1,
                "start": 1,
              },
              "type": "Span",
            },
            "tag": "VariableTerm",
            "variable": "x",
          },
          "span": Object {
            "column": Object {
              "end": 17,
              "start": 9,
            },
            "line": Object {
              "end": 1,
              "start": 1,
            },
            "type": "Span",
          },
          "tag": "LambdaTerm",
        },
        "span": Object {
          "column": Object {
            "end": 17,
            "start": 5,
          },
          "line": Object {
            "end": 1,
            "start": 1,
          },
          "type": "Span",
        },
        "tag": "VariableDeclaration",
        "variable": "f",
      },
    ],
    "span": Object {
      "column": Object {
        "end": 25,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "LetTerm",
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "example1": example1 1`] = `
Object {
  "term": Object {
    "body": Object {
      "block": Object {
        "span": Object {
          "column": Object {
            "end": 2,
            "start": 7,
          },
          "line": Object {
            "end": 15,
            "start": 10,
          },
          "type": "Span",
        },
        "statements": Array [
          Object {
            "lhs": "baz",
            "rhs": Object {
              "args": Array [
                Object {
                  "span": Object {
                    "column": Object {
                      "end": 15,
                      "start": 12,
                    },
                    "line": Object {
                      "end": 11,
                      "start": 11,
                    },
                    "type": "Span",
                  },
                  "tag": "VariableTerm",
                  "variable": "bar",
                },
              ],
              "func": Object {
                "span": Object {
                  "column": Object {
                    "end": 11,
                    "start": 9,
                  },
                  "line": Object {
                    "end": 11,
                    "start": 11,
                  },
                  "type": "Span",
                },
                "tag": "VariableTerm",
                "variable": "id",
              },
              "span": Object {
                "column": Object {
                  "end": 16,
                  "start": 9,
                },
                "line": Object {
                  "end": 11,
                  "start": 11,
                },
                "type": "Span",
              },
              "tag": "CallTerm",
            },
            "span": Object {
              "column": Object {
                "end": 16,
                "start": 3,
              },
              "line": Object {
                "end": 11,
                "start": 11,
              },
              "type": "Span",
            },
            "tag": "AssignmentStatement",
          },
          Object {
            "args": Array [
              Object {
                "span": Object {
                  "column": Object {
                    "end": 15,
                    "start": 12,
                  },
                  "line": Object {
                    "end": 12,
                    "start": 12,
                  },
                  "type": "Span",
                },
                "tag": "VariableTerm",
                "variable": "baz",
              },
              Object {
                "span": Object {
                  "column": Object {
                    "end": 20,
                    "start": 17,
                  },
                  "line": Object {
                    "end": 12,
                    "start": 12,
                  },
                  "type": "Span",
                },
                "tag": "VariableTerm",
                "variable": "bar",
              },
            ],
            "func": Object {
              "span": Object {
                "column": Object {
                  "end": 11,
                  "start": 3,
                },
                "line": Object {
                  "end": 12,
                  "start": 12,
                },
                "type": "Span",
              },
              "tag": "VariableTerm",
              "variable": "constant",
            },
            "span": Object {
              "column": Object {
                "end": 21,
                "start": 3,
              },
              "line": Object {
                "end": 12,
                "start": 12,
              },
              "type": "Span",
            },
            "tag": "CallStatement",
          },
          Object {
            "lhs": "clientName",
            "rhs": Object {
              "member": "name",
              "object": Object {
                "args": Array [
                  Object {
                    "span": Object {
                      "column": Object {
                        "end": 29,
                        "start": 26,
                      },
                      "line": Object {
                        "end": 13,
                        "start": 13,
                      },
                      "type": "Span",
                    },
                    "tag": "VariableTerm",
                    "variable": "baz",
                  },
                ],
                "func": Object {
                  "span": Object {
                    "column": Object {
                      "end": 25,
                      "start": 16,
                    },
                    "line": Object {
                      "end": 13,
                      "start": 13,
                    },
                    "type": "Span",
                  },
                  "tag": "VariableTerm",
                  "variable": "createFoo",
                },
                "span": Object {
                  "column": Object {
                    "end": 30,
                    "start": 16,
                  },
                  "line": Object {
                    "end": 13,
                    "start": 13,
                  },
                  "type": "Span",
                },
                "tag": "CallTerm",
              },
              "span": Object {
                "column": Object {
                  "end": 35,
                  "start": 16,
                },
                "line": Object {
                  "end": 13,
                  "start": 13,
                },
                "type": "Span",
              },
              "tag": "MemberAccessTerm",
            },
            "span": Object {
              "column": Object {
                "end": 35,
                "start": 3,
              },
              "line": Object {
                "end": 13,
                "start": 13,
              },
              "type": "Span",
            },
            "tag": "AssignmentStatement",
          },
          Object {
            "result": Object {
              "span": Object {
                "column": Object {
                  "end": 20,
                  "start": 10,
                },
                "line": Object {
                  "end": 14,
                  "start": 14,
                },
                "type": "Span",
              },
              "tag": "VariableTerm",
              "variable": "clientName",
            },
            "span": Object {
              "column": Object {
                "end": 20,
                "start": 3,
              },
              "line": Object {
                "end": 14,
                "start": 14,
              },
              "type": "Span",
            },
            "tag": "ReturnStatement",
          },
        ],
        "tag": "BlockStatement",
      },
      "span": Object {
        "column": Object {
          "end": 2,
          "start": 4,
        },
        "line": Object {
          "end": 15,
          "start": 10,
        },
        "type": "Span",
      },
      "tag": "DoTerm",
    },
    "declarations": Array [
      Object {
        "body": Object {
          "args": Array [
            "x",
          ],
          "body": Object {
            "span": Object {
              "column": Object {
                "end": 16,
                "start": 15,
              },
              "line": Object {
                "end": 3,
                "start": 3,
              },
              "type": "Span",
            },
            "tag": "VariableTerm",
            "variable": "x",
          },
          "span": Object {
            "column": Object {
              "end": 16,
              "start": 8,
            },
            "line": Object {
              "end": 3,
              "start": 3,
            },
            "type": "Span",
          },
          "tag": "LambdaTerm",
        },
        "span": Object {
          "column": Object {
            "end": 16,
            "start": 3,
          },
          "line": Object {
            "end": 3,
            "start": 3,
          },
          "type": "Span",
        },
        "tag": "VariableDeclaration",
        "variable": "id",
      },
      Object {
        "body": Object {
          "args": Array [
            "a",
            "b",
          ],
          "body": Object {
            "span": Object {
              "column": Object {
                "end": 25,
                "start": 24,
              },
              "line": Object {
                "end": 4,
                "start": 4,
              },
              "type": "Span",
            },
            "tag": "VariableTerm",
            "variable": "a",
          },
          "span": Object {
            "column": Object {
              "end": 25,
              "start": 14,
            },
            "line": Object {
              "end": 4,
              "start": 4,
            },
            "type": "Span",
          },
          "tag": "LambdaTerm",
        },
        "span": Object {
          "column": Object {
            "end": 25,
            "start": 3,
          },
          "line": Object {
            "end": 4,
            "start": 4,
          },
          "type": "Span",
        },
        "tag": "VariableDeclaration",
        "variable": "constant",
      },
      Object {
        "body": Object {
          "args": Array [
            "x",
          ],
          "body": Object {
            "entries": Object {
              "name": Object {
                "elements": Array [
                  Object {
                    "span": Object {
                      "column": Object {
                        "end": 15,
                        "start": 12,
                      },
                      "line": Object {
                        "end": 6,
                        "start": 6,
                      },
                      "type": "Span",
                    },
                    "tag": "ValueTerm",
                    "value": Object {
                      "span": Object {
                        "column": Object {
                          "end": 15,
                          "start": 12,
                        },
                        "line": Object {
                          "end": 6,
                          "start": 6,
                        },
                        "type": "Span",
                      },
                      "str": "F",
                      "tag": "StringValue",
                    },
                  },
                  Object {
                    "span": Object {
                      "column": Object {
                        "end": 20,
                        "start": 17,
                      },
                      "line": Object {
                        "end": 6,
                        "start": 6,
                      },
                      "type": "Span",
                    },
                    "tag": "ValueTerm",
                    "value": Object {
                      "span": Object {
                        "column": Object {
                          "end": 20,
                          "start": 17,
                        },
                        "line": Object {
                          "end": 6,
                          "start": 6,
                        },
                        "type": "Span",
                      },
                      "str": "o",
                      "tag": "StringValue",
                    },
                  },
                  Object {
                    "span": Object {
                      "column": Object {
                        "end": 25,
                        "start": 22,
                      },
                      "line": Object {
                        "end": 6,
                        "start": 6,
                      },
                      "type": "Span",
                    },
                    "tag": "ValueTerm",
                    "value": Object {
                      "span": Object {
                        "column": Object {
                          "end": 25,
                          "start": 22,
                        },
                        "line": Object {
                          "end": 6,
                          "start": 6,
                        },
                        "type": "Span",
                      },
                      "str": "o",
                      "tag": "StringValue",
                    },
                  },
                  Object {
                    "span": Object {
                      "column": Object {
                        "end": 30,
                        "start": 27,
                      },
                      "line": Object {
                        "end": 6,
                        "start": 6,
                      },
                      "type": "Span",
                    },
                    "tag": "ValueTerm",
                    "value": Object {
                      "span": Object {
                        "column": Object {
                          "end": 30,
                          "start": 27,
                        },
                        "line": Object {
                          "end": 6,
                          "start": 6,
                        },
                        "type": "Span",
                      },
                      "str": "b",
                      "tag": "StringValue",
                    },
                  },
                  Object {
                    "span": Object {
                      "column": Object {
                        "end": 35,
                        "start": 32,
                      },
                      "line": Object {
                        "end": 6,
                        "start": 6,
                      },
                      "type": "Span",
                    },
                    "tag": "ValueTerm",
                    "value": Object {
                      "span": Object {
                        "column": Object {
                          "end": 35,
                          "start": 32,
                        },
                        "line": Object {
                          "end": 6,
                          "start": 6,
                        },
                        "type": "Span",
                      },
                      "str": "a",
                      "tag": "StringValue",
                    },
                  },
                  Object {
                    "span": Object {
                      "column": Object {
                        "end": 40,
                        "start": 37,
                      },
                      "line": Object {
                        "end": 6,
                        "start": 6,
                      },
                      "type": "Span",
                    },
                    "tag": "ValueTerm",
                    "value": Object {
                      "span": Object {
                        "column": Object {
                          "end": 40,
                          "start": 37,
                        },
                        "line": Object {
                          "end": 6,
                          "start": 6,
                        },
                        "type": "Span",
                      },
                      "str": "r",
                      "tag": "StringValue",
                    },
                  },
                ],
                "span": Object {
                  "column": Object {
                    "end": 41,
                    "start": 11,
                  },
                  "line": Object {
                    "end": 6,
                    "start": 6,
                  },
                  "type": "Span",
                },
                "tag": "ArrayTerm",
              },
              "value": Object {
                "span": Object {
                  "column": Object {
                    "end": 13,
                    "start": 12,
                  },
                  "line": Object {
                    "end": 7,
                    "start": 7,
                  },
                  "type": "Span",
                },
                "tag": "VariableTerm",
                "variable": "x",
              },
            },
            "span": Object {
              "column": Object {
                "end": 4,
                "start": 22,
              },
              "line": Object {
                "end": 8,
                "start": 5,
              },
              "type": "Span",
            },
            "tag": "ObjectTerm",
          },
          "span": Object {
            "column": Object {
              "end": 4,
              "start": 15,
            },
            "line": Object {
              "end": 8,
              "start": 5,
            },
            "type": "Span",
          },
          "tag": "LambdaTerm",
        },
        "span": Object {
          "column": Object {
            "end": 4,
            "start": 3,
          },
          "line": Object {
            "end": 8,
            "start": 5,
          },
          "type": "Span",
        },
        "tag": "VariableDeclaration",
        "variable": "createFoo",
      },
      Object {
        "body": Object {
          "span": Object {
            "column": Object {
              "end": 12,
              "start": 9,
            },
            "line": Object {
              "end": 9,
              "start": 9,
            },
            "type": "Span",
          },
          "tag": "ValueTerm",
          "value": Object {
            "num": 999,
            "span": Object {
              "column": Object {
                "end": 12,
                "start": 9,
              },
              "line": Object {
                "end": 9,
                "start": 9,
              },
              "type": "Span",
            },
            "tag": "NumberValue",
          },
        },
        "span": Object {
          "column": Object {
            "end": 12,
            "start": 3,
          },
          "line": Object {
            "end": 9,
            "start": 9,
          },
          "type": "Span",
        },
        "tag": "VariableDeclaration",
        "variable": "bar",
      },
    ],
    "span": Object {
      "column": Object {
        "end": 2,
        "start": 1,
      },
      "line": Object {
        "end": 15,
        "start": 2,
      },
      "type": "Span",
    },
    "tag": "LetTerm",
  },
  "warnings": Array [
    [Error: Ambiguous grammar encountered! Resolving first result.],
  ],
}
`;

exports[`Parser Golden Test Suite Parse example "lambda1": lambda1 1`] = `
Object {
  "term": Object {
    "args": Array [
      "x",
    ],
    "body": Object {
      "span": Object {
        "column": Object {
          "end": 9,
          "start": 8,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "VariableTerm",
      "variable": "x",
    },
    "span": Object {
      "column": Object {
        "end": 9,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "LambdaTerm",
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "let1": let1 1`] = `
Object {
  "term": Object {
    "body": Object {
      "span": Object {
        "column": Object {
          "end": 15,
          "start": 14,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "VariableTerm",
      "variable": "x",
    },
    "declarations": Array [
      Object {
        "body": Object {
          "span": Object {
            "column": Object {
              "end": 10,
              "start": 9,
            },
            "line": Object {
              "end": 1,
              "start": 1,
            },
            "type": "Span",
          },
          "tag": "ValueTerm",
          "value": Object {
            "num": 1,
            "span": Object {
              "column": Object {
                "end": 10,
                "start": 9,
              },
              "line": Object {
                "end": 1,
                "start": 1,
              },
              "type": "Span",
            },
            "tag": "NumberValue",
          },
        },
        "span": Object {
          "column": Object {
            "end": 10,
            "start": 5,
          },
          "line": Object {
            "end": 1,
            "start": 1,
          },
          "type": "Span",
        },
        "tag": "VariableDeclaration",
        "variable": "x",
      },
    ],
    "span": Object {
      "column": Object {
        "end": 15,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "LetTerm",
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "let2": let2 1`] = `
Object {
  "term": Object {
    "body": Object {
      "span": Object {
        "column": Object {
          "end": 16,
          "start": 15,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "VariableTerm",
      "variable": "x",
    },
    "declarations": Array [
      Object {
        "body": Object {
          "span": Object {
            "column": Object {
              "end": 10,
              "start": 9,
            },
            "line": Object {
              "end": 1,
              "start": 1,
            },
            "type": "Span",
          },
          "tag": "ValueTerm",
          "value": Object {
            "num": 1,
            "span": Object {
              "column": Object {
                "end": 10,
                "start": 9,
              },
              "line": Object {
                "end": 1,
                "start": 1,
              },
              "type": "Span",
            },
            "tag": "NumberValue",
          },
        },
        "span": Object {
          "column": Object {
            "end": 10,
            "start": 5,
          },
          "line": Object {
            "end": 1,
            "start": 1,
          },
          "type": "Span",
        },
        "tag": "VariableDeclaration",
        "variable": "x",
      },
    ],
    "span": Object {
      "column": Object {
        "end": 16,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "LetTerm",
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "let3": let3 1`] = `
Object {
  "term": Object {
    "body": Object {
      "elements": Array [
        Object {
          "span": Object {
            "column": Object {
              "end": 24,
              "start": 23,
            },
            "line": Object {
              "end": 1,
              "start": 1,
            },
            "type": "Span",
          },
          "tag": "VariableTerm",
          "variable": "x",
        },
        Object {
          "span": Object {
            "column": Object {
              "end": 27,
              "start": 26,
            },
            "line": Object {
              "end": 1,
              "start": 1,
            },
            "type": "Span",
          },
          "tag": "VariableTerm",
          "variable": "y",
        },
      ],
      "span": Object {
        "column": Object {
          "end": 28,
          "start": 22,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "ArrayTerm",
    },
    "declarations": Array [
      Object {
        "body": Object {
          "span": Object {
            "column": Object {
              "end": 10,
              "start": 9,
            },
            "line": Object {
              "end": 1,
              "start": 1,
            },
            "type": "Span",
          },
          "tag": "ValueTerm",
          "value": Object {
            "num": 1,
            "span": Object {
              "column": Object {
                "end": 10,
                "start": 9,
              },
              "line": Object {
                "end": 1,
                "start": 1,
              },
              "type": "Span",
            },
            "tag": "NumberValue",
          },
        },
        "span": Object {
          "column": Object {
            "end": 10,
            "start": 5,
          },
          "line": Object {
            "end": 1,
            "start": 1,
          },
          "type": "Span",
        },
        "tag": "VariableDeclaration",
        "variable": "x",
      },
      Object {
        "body": Object {
          "span": Object {
            "column": Object {
              "end": 17,
              "start": 16,
            },
            "line": Object {
              "end": 1,
              "start": 1,
            },
            "type": "Span",
          },
          "tag": "ValueTerm",
          "value": Object {
            "num": 2,
            "span": Object {
              "column": Object {
                "end": 17,
                "start": 16,
              },
              "line": Object {
                "end": 1,
                "start": 1,
              },
              "type": "Span",
            },
            "tag": "NumberValue",
          },
        },
        "span": Object {
          "column": Object {
            "end": 17,
            "start": 12,
          },
          "line": Object {
            "end": 1,
            "start": 1,
          },
          "type": "Span",
        },
        "tag": "VariableDeclaration",
        "variable": "y",
      },
    ],
    "span": Object {
      "column": Object {
        "end": 28,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "LetTerm",
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "null1": null1 1`] = `
Object {
  "errors": Array [
    [Error: No parse results were recovered.],
  ],
  "term": null,
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "null2": null2 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 5,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "span": Object {
        "column": Object {
          "end": 5,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "NullValue",
    },
  },
  "warnings": Array [
    [Error: Ambiguous grammar encountered! Resolving first result.],
  ],
}
`;

exports[`Parser Golden Test Suite Parse example "number1": number1 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 2,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "num": 0,
      "span": Object {
        "column": Object {
          "end": 2,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "NumberValue",
    },
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "number2": number2 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 11,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "num": 1234567890,
      "span": Object {
        "column": Object {
          "end": 11,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "NumberValue",
    },
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "number3": number3 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 22,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "num": 1234567890.0123458,
      "span": Object {
        "column": Object {
          "end": 22,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "tag": "NumberValue",
    },
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "string1": string1 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 3,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "span": Object {
        "column": Object {
          "end": 3,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "str": "",
      "tag": "StringValue",
    },
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "string2": string2 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 3,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "span": Object {
        "column": Object {
          "end": 3,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "str": "",
      "tag": "StringValue",
    },
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "string3": string3 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 15,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "span": Object {
        "column": Object {
          "end": 15,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "str": "Hello world!",
      "tag": "StringValue",
    },
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "string4": string4 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 15,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "span": Object {
        "column": Object {
          "end": 15,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "str": "Hello World!",
      "tag": "StringValue",
    },
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "string5": string5 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 47,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "span": Object {
        "column": Object {
          "end": 47,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "str": "The quick brown fox jumps over the lazy dog.",
      "tag": "StringValue",
    },
  },
  "warnings": Array [],
}
`;

exports[`Parser Golden Test Suite Parse example "string6": string6 1`] = `
Object {
  "term": Object {
    "span": Object {
      "column": Object {
        "end": 47,
        "start": 1,
      },
      "line": Object {
        "end": 1,
        "start": 1,
      },
      "type": "Span",
    },
    "tag": "ValueTerm",
    "value": Object {
      "span": Object {
        "column": Object {
          "end": 47,
          "start": 1,
        },
        "line": Object {
          "end": 1,
          "start": 1,
        },
        "type": "Span",
      },
      "str": "The quick brown fox jumps over the lazy dog.",
      "tag": "StringValue",
    },
  },
  "warnings": Array [],
}
`;
